### PR TITLE
XDS Server: introducing snapshot cache

### PR DESCRIPTION
**Pull request by eduser25**
Go-control-plane includes an XDS server implementation that is interacted
through a map-like mechanisms that helps decouple generating proxy configuration
from its distribution.

https://github.com/envoyproxy/go-control-plane/blob/main/docs/cache/Snapshot.md

This commit adds support for initializing OSM control plane to use such mechanism, behind
a feature flag.

Using the cache currently suffers from some problems (Some e2e work, Some are flacky, most will fail).
More work is to be put on top of this implementation to debug these issues. This commit introduces merely
the foundations to keep things short.

Meaningful changes:
- New featureflag added: `enable-snapshotcache`
- `server.go` will initialize the XDS server with or without the cache when enabled.
- Some changes were made to `ads/response.go` to accomodate both modes working.
Namely, it will now `get` the resources for a vertical, and later decide what to do with
them (either send to a connected proxy, or store them in the cache, depending if the cache
is enabled or not).
- `stream.go` was before the routine receiving events and triggering envoy re-configs. When
working on snapshot-cache, such routines do not exist, so a new routine to mimic the same
current design and behavior has been added (`snapshot_cache_stream.go`). Its purpose is to
listen to broadcast events from dispatcher, and issue a reconfiguration for all proxies when
a broadcast update is sent from dispatcher.

TODOs:
- Proxy objects are not created anymore in snapshot-cache based. Follow up discussion if
we want to have them or not is to happen.
- Zerolog doesn't implement Golang's logging interface and can't be used/passed to log
cache internals.
- Additional context validation (certificate) can happen on cache callback context.
- We have no way to infer the `kind` of proxy from a Pod. Right now it's temporarily
hardcoded as `sidecar`.

Signed-off-by: Eduard Serra <eduser25@gmail.com>

| New Functionality          | [x] |
| Control Plane              | [x] |

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?
No
1. Is this a breaking change?
No

### CODE REVIEW

File: pkg/envoy/ads/snapshot_cache_stream.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		return nil, errors.Errorf("UUID not found for pod %s/%s, not a mesh pod", pod.Namespace, pod.Name)
+	}
+
+	serviceAccount = pod.Spec.ServiceAccountName
+	namespace = pod.Namespace
+
+	// construct CN for this pod/proxy
+	// TODO: Infer proxy type from Pod
+	commonName := strings.Join([]string{uuid, "sidecar", serviceAccount, namespace}, constants.DomainDelimiter)
+	tempProxy := envoy.NewProxy(certificate.CommonName(commonName), "NoSerial", &net.IPAddr{IP: net.IPv4zero})
```
Comment (eduser25) [inline pkg/envoy/ads/snapshot_cache_stream.go:71]: ?????

File: pkg/envoy/ads/response.go [hunk 2]
```diff
@@ -1,36 +1,47 @@
...
 	"github.com/envoyproxy/go-control-plane/pkg/cache/v3"
 	"github.com/golang/protobuf/ptypes"
 
 	"github.com/openservicemesh/osm/pkg/configurator"
 	"github.com/openservicemesh/osm/pkg/envoy"
 )
 
-// Wrapper to create and send a discovery response to an envoy server
-func (s *Server) sendTypeResponse(typeURI envoy.TypeURI, proxy *envoy.Proxy, server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer, req *xds_discovery.DiscoveryRequest, cfg configurator.Configurator) error {
+// getTypeResource invokes the XDS vertical to get the resources to be sent associated to the Type and the Proxy passed by parametre
```
Comment (shashankram) [inline pkg/envoy/ads/response.go:18]:
```suggestion
// getTypeResource invokes the XDS handler (LDS, CDS etc.) to respond to the XDS request containing the requests' type and associated resources
```

File: pkg/envoy/ads/response.go [hunk 3]
```diff
@@ -1,36 +1,47 @@
...
+func (s *Server) getTypeResources(proxy *envoy.Proxy, request *xds_discovery.DiscoveryRequest) ([]types.Resource, error) {
 	// Tracks the success of this TypeURI response operation; accounts also for receipt on envoy server side
 	startedAt := time.Now()
-	log.Trace().Msgf("[%s] Creating response for proxy with SerialNumber=%s on Pod with UID=%s", typeURI.Short(), proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+	typeURI := envoy.TypeURI(request.TypeUrl)
+	log.Trace().Msgf("[%s] Getting resources for proxy with SerialNumber=%s on Pod with UID=%s", typeURI.Short(), proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
```
Comment (shashankram) [inline pkg/envoy/ads/response.go:23]: > proxy with SerialNumber=%s on Pod with UID=%s

Every time I see this log 👀

File: pkg/envoy/ads/response.go [hunk 4]
```diff
@@ -76,56 +110,68 @@ func (s *Server) sendResponse(proxy *envoy.Proxy, server *xds_discovery.Aggregat
...
-func (s *Server) newAggregatedDiscoveryResponse(proxy *envoy.Proxy, request *xds_discovery.DiscoveryRequest, cfg configurator.Configurator) (*xds_discovery.DiscoveryResponse, error) {
-	typeURL := envoy.TypeURI(request.TypeUrl)
-	handler, ok := s.xdsHandlers[typeURL]
-	if !ok {
-		log.Error().Msgf("Responder for TypeUrl %s is not implemented", request.TypeUrl)
-		return nil, errUnknownTypeURL
+// RecordFullSnapshot stores a group of resources as a new Snapshot with a new version in the cache.
+// It also runs a consistency check on the snapshot (will warn if there are missing resources referenced in
+// the snapshot)
+func (s *Server) RecordFullSnapshot(proxy *envoy.Proxy, snapshotResources map[envoy.TypeURI][]types.Resource) error {
```
Comment (shashankram) [inline pkg/envoy/ads/response.go:116]: Consider moving this to `ads/cache.go`?

File: pkg/envoy/ads/server.go [hunk 5]
```diff
@@ -50,6 +53,9 @@ func NewADSServer(meshCatalog catalog.MeshCataloger, proxyRegistry *registry.Pro
 		xdsLog:         make(map[certificate.CommonName]map[envoy.TypeURI][]time.Time),
 		workqueues:     workerpool.NewWorkerPool(workerPoolSize),
 		kubecontroller: kubecontroller,
+		cacheEnabled:   featureflags.IsSnapshotCacheEnabled(),
+		configVerMutex: sync.Mutex{},
```
Comment (shashankram) [inline pkg/envoy/ads/server.go:57]: nil value default?

File: pkg/envoy/ads/server.go [hunk 6]
```diff
@@ -70,8 +76,24 @@ func (s *Server) Start(ctx context.Context, cancel context.CancelFunc, port int,
 		return err
 	}
 
-	xds_discovery.RegisterAggregatedDiscoveryServiceServer(grpcServer, s)
+	if s.cacheEnabled {
+		// TODO: Zerolog can't be passed as snapshot cache internal logger as it doesn't implement golang's logger interface
+		s.ch = cachev3.NewSnapshotCache(false, cachev3.IDHash{}, nil)
```
Comment (shashankram) [inline pkg/envoy/ads/server.go:81]: I got the comment, but correlating it to the code is missing.

File: pkg/envoy/ads/snapshot_cache_stream.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	if !uuidFound {
+		return nil, errors.Errorf("UUID not found for pod %s/%s, not a mesh pod", pod.Namespace, pod.Name)
+	}
+
+	serviceAccount = pod.Spec.ServiceAccountName
+	namespace = pod.Namespace
+
+	// construct CN for this pod/proxy
+	// TODO: Infer proxy type from Pod
+	commonName := strings.Join([]string{uuid, "sidecar", serviceAccount, namespace}, constants.DomainDelimiter)
```
[the remark this answers is not in the document]
Comment (eduser25) [reply pkg/envoy/ads/snapshot_cache_stream.go:70]: very likely, let me see

File: cmd/osm-controller/osm-controller.go [hunk 8]
```diff
@@ -107,6 +107,7 @@ func init() {
 	flags.BoolVar(&optionalFeatures.WASMStats, "stats-wasm-experimental", false, "Enable a WebAssembly module that generates additional Envoy statistics")
 	flags.BoolVar(&optionalFeatures.EgressPolicy, "enable-egress-policy", false, "Enable OSM's Egress policy API")
 	flags.BoolVar(&optionalFeatures.MulticlusterMode, "enable-multicluster", false, "Enable multicluster mode in OSM")
+	flags.BoolVar(&optionalFeatures.SnapshotCache, "enable-snapshotcache", false, "Enable snapshot cache XDS mode in OSM")
```
Comment (snehachhabria) [inline cmd/osm-controller/osm-controller.go:110]: FYI, we are planning to remove feature flags and move these into the mesh config : https://github.com/openservicemesh/osm/pull/3536 . You might want to rebase this once the former is merged. 

cc @jaellio

File: pkg/envoy/ads/snapshot_cache_stream.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+		s.workqueues.AddJob(&job)
+	}
+}
+
+// GetProxyFromPod infers and creatse a Proxy data structure from a Pod.
+// This is a temporary workaround as proxy is required and expected in any vertical call to XDS,
+// however snapshotcache has no need to provide visibility on proxies whatsoever.
+// All verticals use the proxy structure to infer the pod later, so the actual only mandatory
+// data for the verticals to be functional is the common name, which links proxy <-> pod
+func GetProxyFromPod(pod *v1.Pod) (*envoy.Proxy, error) {
```
Comment (snehachhabria) [inline pkg/envoy/ads/snapshot_cache_stream.go:55]: Could we add a unit test for this ?

File: pkg/envoy/ads/response.go [hunk 10]
```diff
@@ -1,36 +1,47 @@
...
+		s.trackXDSLog(proxy.GetCertificateCommonName(), typeURI)
+	}
+
+	// Invoke XDS handler
+	resources, err := handler(s.catalog, proxy, request, s.cfg, s.certManager, s.proxyRegistry)
+	if err != nil {
+		log.Error().Err(err).Msgf("Handler errored TypeURL: %s, proxy: %s", typeURI.Short(), proxy.GetCertificateSerialNumber())
 		xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, false)
-		return err
+		return nil, errCreatingResponse
```
[the remark this answers is not in the document]
Comment (eduser25) [reply pkg/envoy/ads/response.go:40]: actually... The errors types logged and returned are different. What we return are generic errors while what we log here is more specific. Not sure how to take this, but I think we'd rather change it in a follow up (we should decide if we want specific logs to be returned, generic logs to be logged, etc).

File: pkg/envoy/ads/snapshot_cache_stream.go (hunk 9 above)
Comment (eduser25) [reply pkg/envoy/ads/snapshot_cache_stream.go:55]: writing :+1:

File: cmd/osm-controller/osm-controller.go (hunk 8 above)
Comment (eduser25) [reply cmd/osm-controller/osm-controller.go:110]: are we racing? :eyes:

**Review by shashankram — APPROVED** [review]

File: cmd/osm-controller/osm-controller.go (hunk 8 above)
Comment (eduser25) [reply cmd/osm-controller/osm-controller.go]: rebased

**Review by snehachhabria — APPROVED** [review]

**Pull request merged**